osutil/sync_dir.go

```diff
@@ -0,0 +1,159 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"syscall"
+)
+
+// FileState describes the expected content and meta data of a single file.
+//
+// FileState is used by the SyncDir()
+type FileState struct {
+	// Content is the entire content that a specific file is expected to have.
+	Content []byte
+	// Mode describes the permissions that the file is expected to have.
+	Mode os.FileMode
+	// UID is the identifier of the user who owns the file (UNIX only).
+	UID uint32
+	// GID is the identifier of the group who owns the file (UNIX only).
+	GID uint32
+}
+
+// SyncDir ensures that directory content matches expectations.
```

Comment:
@mvo5 suggested EnsureDirState to avoid the rsync feeling of DirSync.. sounds good.